app/src/org/commcare/activities/GlobalPrivilegeClaimingActivity.java

```diff
@@ -101,72 +105,45 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
         switch(requestCode) {
             case BARCODE_CAPTURE:
                 if (resultCode == RESULT_OK) {
-                    String[] fields = processScanResult(data.getStringExtra("SCAN_RESULT"));
-                    if (fields == null) {
-                        privilegeClaimFailed();
-                    } else {
-                        String flag = fields[0];
-                        String username = fields[1];
-                        String signature = fields[2];
-                        if (checkProperFormAndAuthenticity(flag, username, signature)) {
-                            GlobalPrivilegesManager.enablePrivilege(flag, username);
-                            refreshUI();
-                        } else {
-                            privilegeClaimFailed();
+                    String scanResult = data.getStringExtra("SCAN_RESULT");
+                    try {
+                        Pair<String, String[]> activatedPrivileges =
+                                new PrivilegesUtility(GlobalConstants.TRUSTED_SOURCE_PUBLIC_KEY).
+                                        processPrivilegePayloadForActivatedPrivileges(scanResult);
+
+                        for(String p : activatedPrivileges.second) {
+                            if (!GlobalPrivilegesManager.allGlobalPrivilegesList.contains(p)) {
```

Comment:
Mm ok thought through it a more and decided it doesn't really change much, since all it would mean is that someone could theoretically have a new privileges working automatically as soon as they performed an upgrade to a version of CommCare containing that privilege (if they had previously scanned a barcode containing it). But I don't know that there's much value in that, and could maybe result in weird/unexpected happenings. So I think fine to leave.